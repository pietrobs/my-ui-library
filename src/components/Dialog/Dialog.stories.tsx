import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Dialog from './Dialog';

export default {
  title: 'Example/Dialog',
  component: Dialog,
  argTypes: {
    closeOnOverlayClick: {
      name: 'Close on overlay click',
      defaultValue: true,
      type: 'boolean',
      description: 'Enable click on overlay to close the Dialog',
    },
    children: {
      name: 'Content',
      description: 'Content of the dialog',
    },
  },
} as ComponentMeta<typeof Dialog>;

const Template: ComponentStory<typeof Dialog> = (args) => <Dialog {...args} />;

export const OverlayClickEnabled = Template.bind({});

OverlayClickEnabled.args = {
  title: 'Dialog title',
  children: 'Dialog content',
  isOpen: true,
  onClose: () => {
    alert('Call onClose function');
  },
  closeOnOverlayClick: true,
};

export const OverlayClickDisabled = Template.bind({});
OverlayClickDisabled.args = {
  title: 'Dialog title',
  children: 'Dialog content',
  isOpen: true,
  onClose: () => {
    alert('Call onClose function');
  },
  closeOnOverlayClick: false,
};

export const ClosedDialog = Template.bind({});
ClosedDialog.args = {
  title: 'Dialog title',
  children: 'Dialog content',
  isOpen: false,
  onClose: () => {
    alert('Call onClose function');
  },
  closeOnOverlayClick: false,
};

const bigText = `
Lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum ipsum
lorem faszipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum 
`;

export const WithBigContent = Template.bind({});
WithBigContent.args = {
  title: 'Dialog title',
  children: bigText,
  isOpen: true,
  onClose: () => {
    alert('Call onClose function');
  },
  closeOnOverlayClick: false,
};

export const WithoutTitle = Template.bind({});
WithoutTitle.args = {
  children: bigText,
  isOpen: true,
  onClose: () => {
    alert('Call onClose function');
  },
  closeOnOverlayClick: false,
};
