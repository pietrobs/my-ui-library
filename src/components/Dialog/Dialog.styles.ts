import styled from 'styled-components';

export const DialogWrapper = styled.div`
  .dialog-overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.5);
    display: flex;
    justify-content: center;
    align-items: flex-end;
    z-index: 102;

    .dialog {
      display: flex;
      flex-direction: column;
      background-color: #fff;
      padding: 32px;
      width: 100%;
      max-height: 80%;

      .dialog-header {
        display: flex;
        justify-content: space-between;
        align-items: center;
        margin-bottom: 20px;

        h1 {
          font-weight: 500;
          font-size: 22px;
          line-height: 30px;
          letter-spacing: -0.032em;
        }

        button {
          background-color: transparent;
          border: none;
          cursor: pointer;
          outline: none;
        }
      }

      .dialog-content {
        overflow-y: auto;
        height: auto;

        font-weight: 400;
        font-size: 18px;
        line-height: 24px;
        letter-spacing: -0.028em;

        ::-webkit-scrollbar {
          display: none;
        }
      }
    }
  }
`;
