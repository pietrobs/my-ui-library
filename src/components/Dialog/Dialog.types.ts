export interface DialogProps {
  title?: string;
  isOpen: boolean;
  onClose: () => void;
  closeOnOverlayClick: boolean;
  children?: React.ReactNode;
}

export interface DialogHeaderProps {
  title?: string;
  onClose: () => void;
}
