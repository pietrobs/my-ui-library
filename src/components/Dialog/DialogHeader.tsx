import React from 'react';
import { DialogHeaderProps } from './Dialog.types';

const DialogHeader: React.FC<DialogHeaderProps> = ({ onClose, title }) => {
  return (
    <div className="dialog-header">
      <h1>{title}</h1>
      <button
        onClick={() => {
          onClose();
        }}
      >
        <svg
          width="16"
          height="16"
          viewBox="0 0 16 16"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path d="M15 1L8 8L15 15" stroke="black" strokeWidth="1.2" />
          <path d="M1 1L8 8L0.999999 15" stroke="black" strokeWidth="1.2" />
        </svg>
      </button>
    </div>
  );
};

export default DialogHeader;
