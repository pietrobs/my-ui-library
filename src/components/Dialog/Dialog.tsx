import React from 'react';
import { useRef } from 'react';
import ReactDOM from 'react-dom';
import useClickAwayListener from '../../utils/useClickAwayListener';
import useKeyPressListener from '../../utils/useKeyPressListener';
import { DialogWrapper } from './Dialog.styles';
import { DialogProps } from './Dialog.types';
import DialogHeader from './DialogHeader';

const Dialog: React.FC<DialogProps> = ({
  children,
  isOpen,
  onClose,
  closeOnOverlayClick,
  title,
}) => {
  const listenerClickAwayRef = useRef<HTMLDivElement>(null);

  useClickAwayListener(
    listenerClickAwayRef,
    closeOnOverlayClick ? onClose : undefined,
  );

  useKeyPressListener('Escape', isOpen ? onClose : undefined);

  if (!isOpen) return null;

  return ReactDOM.createPortal(
    <DialogWrapper>
      <div className="dialog-overlay" data-testid="dialog-overlay">
        <div className="dialog" ref={listenerClickAwayRef}>
          <DialogHeader title={title} onClose={onClose} />
          <div className="dialog-content">{children}</div>
        </div>
      </div>
    </DialogWrapper>,
    document.body,
  );
};

export default Dialog;
