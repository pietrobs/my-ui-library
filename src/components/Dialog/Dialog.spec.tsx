import React from 'react';
import { describe, test, expect, it, jest } from '@jest/globals';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';

import Dialog from './Dialog';

const dialogTitle = 'Title';
const dialogContent = 'Dialog content';

const MockedComponent = ({
  closeOnOverlayClick = true,
  onClose = () => {},
  isOpen = true,
}: {
  closeOnOverlayClick?: boolean;
  onClose?: () => void;
  isOpen?: boolean;
}) => (
  <Dialog
    closeOnOverlayClick={closeOnOverlayClick}
    isOpen={isOpen}
    onClose={onClose}
    title={dialogTitle}
  >
    {dialogContent}
  </Dialog>
);

describe('Dialog Component Openned', () => {
  test('Should match snapshot', () => {
    ReactDOM.createPortal = jest.fn((element, node) => {
      return element;
    }) as any;

    const tree = renderer.create(<MockedComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Title is renderized correctly', () => {
    render(<MockedComponent />);

    const title = screen.queryByText(dialogTitle);

    expect(title?.textContent).toBe(dialogTitle);
  });

  test('Content is renderized correctly', () => {
    render(<MockedComponent />);

    const title = screen.queryByText(dialogContent);

    expect(title?.textContent).toBe(dialogContent);
  });

  test('Clicking on overlay close the dialog', () => {
    const onClose = jest.fn();
    render(<MockedComponent closeOnOverlayClick onClose={onClose} />);
    const overlay = screen.getByTestId('dialog-overlay');

    fireEvent.click(overlay);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test('Clicking on close button close the dialog', () => {
    const onClose = jest.fn();
    render(<MockedComponent closeOnOverlayClick onClose={onClose} />);
    const closeButton = screen.getByRole('button');

    fireEvent.click(closeButton);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test('Pressing escape close the dialog', () => {
    const onClose = jest.fn();
    render(<MockedComponent closeOnOverlayClick onClose={onClose} />);

    fireEvent.keyUp(document, { key: 'Escape', code: 'Escape' });

    expect(onClose).toHaveBeenCalledTimes(1);
    cleanup();
  });

  test('Clicking on overlay does not close the dialog', () => {
    const onClose = jest.fn();
    render(<MockedComponent closeOnOverlayClick={false} onClose={onClose} />);
    const overlay = screen.getByTestId('dialog-overlay');

    fireEvent.click(overlay);

    expect(onClose).toHaveBeenCalledTimes(0);
  });

  test('Clicking on close button close the dialog', () => {
    const onClose = jest.fn();
    render(<MockedComponent closeOnOverlayClick={false} onClose={onClose} />);
    const closeButton = screen.getByRole('button');

    fireEvent.click(closeButton);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  test('Pressing escape close the dialog', () => {
    const onClose = jest.fn();
    render(<MockedComponent closeOnOverlayClick={false} onClose={onClose} />);

    fireEvent.keyUp(document, { key: 'Escape', code: 'Escape' });

    expect(onClose).toHaveBeenCalledTimes(1);
  });
});

describe('Dialog Component Closed', () => {
  test('Title is renderized correctly', () => {
    const { rerender } = render(<MockedComponent isOpen={false} />);

    let title = screen.queryByText(dialogTitle);

    expect(title?.textContent).toBeFalsy();

    rerender(<MockedComponent isOpen />);

    title = screen.queryByText(dialogTitle);

    expect(title?.textContent).toBe(dialogTitle);
  });
});
