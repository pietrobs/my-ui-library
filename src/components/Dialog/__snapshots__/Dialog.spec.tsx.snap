// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog Component Openned Should match snapshot 1`] = `
<div
  className="sc-bcXHqe dVqXFG"
>
  <div
    className="dialog-overlay"
    data-testid="dialog-overlay"
  >
    <div
      className="dialog"
    >
      <div
        className="dialog-header"
      >
        <h1>
          Title
        </h1>
        <button
          onClick={[Function]}
        >
          <svg
            fill="none"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 1L8 8L15 15"
              stroke="black"
              strokeWidth="1.2"
            />
            <path
              d="M1 1L8 8L0.999999 15"
              stroke="black"
              strokeWidth="1.2"
            />
          </svg>
        </button>
      </div>
      <div
        className="dialog-content"
      >
        Dialog content
      </div>
    </div>
  </div>
</div>
`;
